File: src/surface/hl-gl/tests/fixtures/gl_ubo_transform.c
/* REAL GL UBO TRANSFORM — a real EGL + GLES3 offscreen program whose MVP lives in a std140 uniform BLOCK
 * (set the UBO way: glGenBuffers + glBufferData(std140 bytes) + glBindBufferBase(GL_UNIFORM_BUFFER, 0, ubo)
 * + glUniformBlockBinding), rasterized on lavapipe. The EXACT-RECT counterpart of gl_ubo.c.
 *
 * Where gl_ubo.c proves the block's bytes merely REACH the shader (right-half red vs left-half blue), this
 * program pins the transform to a KNOWN on-screen rectangle: the MVP both SCALES (0.5x, 0.5x) AND TRANSLATES
 * (+0.25 in x), so a full-NDC quad [-1,1]^2 lands at NDC x in [-0.25, 0.75], y in [-0.5, 0.5]. On a 64x64
 * target that is pixel columns [24, 56) x rows [16, 48) — an off-center rectangle. We sample INSIDE (green
 * fill from uColor) and just OUTSIDE all four edges (blue clear): a direct detector that the shim binds the
 * glBindBufferBase'd UBO's std140 bytes at binding 0 (not a synthesized flat/zero block). A wrong route
 * either collapses the geometry (mvp = 0 -> nothing) or misplaces it (identity -> full frame), and either
 * moves at least one of the five sample pixels off its expected color.
 *
 * The x translate makes the rect asymmetric in X (so a swapped/zeroed translation column is visible); the
 * y transform is kept symmetric so the check is independent of readback y-orientation. uColor (also from the
 * block) paints the fill GREEN, so a correct read shows in BOTH geometry position AND pixel color.
 *
 * No hl-specific calls, no vendor headers: the GLES3/EGL ABI is self-declared, and
 * LD_LIBRARY_PATH=~/.hl/gl/<arch> makes libEGL.so.1/libGLESv2.so.2 OURS. Every GL call lowers to hl-GPU IR
 * shipped over $HL_GPU_EXEC to the host WgpuExecutor on lavapipe. Prints "GL_UBOXFORM_OK" on success. */
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef int32_t EGLint;
typedef unsigned int EGLBoolean, EGLenum, GLenum, GLbitfield, GLuint;
typedef int GLint, GLsizei;
typedef float GLfloat;
typedef char GLchar;
typedef void *EGLDisplay, *EGLConfig, *EGLContext, *EGLSurface, *EGLNativeDisplayType;

extern EGLDisplay eglGetDisplay(EGLNativeDisplayType);
extern EGLBoolean eglInitialize(EGLDisplay, EGLint *, EGLint *);
extern EGLBoolean eglChooseConfig(EGLDisplay, const EGLint *, EGLConfig *, EGLint, EGLint *);
extern EGLBoolean eglBindAPI(EGLenum);
extern EGLContext eglCreateContext(EGLDisplay, EGLConfig, EGLContext, const EGLint *);
extern EGLSurface eglCreatePbufferSurface(EGLDisplay, EGLConfig, const EGLint *);
extern EGLBoolean eglMakeCurrent(EGLDisplay, EGLSurface, EGLSurface, EGLContext);

extern GLuint glCreateShader(GLenum);
extern void glShaderSource(GLuint, GLsizei, const GLchar *const *, const GLint *);
extern void glCompileShader(GLuint);
extern GLuint glCreateProgram(void);
extern void glAttachShader(GLuint, GLuint);
extern void glLinkProgram(GLuint);
extern void glUseProgram(GLuint);
extern GLint glGetAttribLocation(GLuint, const GLchar *);
extern GLuint glGetUniformBlockIndex(GLuint, const GLchar *);
extern void glUniformBlockBinding(GLuint, GLuint, GLuint);
extern void glGenBuffers(GLsizei, GLuint *);
extern void glBindBuffer(GLenum, GLuint);
extern void glBindBufferBase(GLenum, GLuint, GLuint);
extern void glBufferData(GLenum, long, const void *, GLenum);
extern void glVertexAttribPointer(GLuint, GLint, GLenum, unsigned char, GLsizei, const void *);
extern void glEnableVertexAttribArray(GLuint);
extern void glClearColor(GLfloat, GLfloat, GLfloat, GLfloat);
extern void glClear(GLbitfield);
extern void glViewport(GLint, GLint, GLsizei, GLsizei);
extern void glDrawArrays(GLenum, GLint, GLsizei);
extern void glReadPixels(GLint, GLint, GLsizei, GLsizei, GLenum, GLenum, void *);
extern GLenum glGetError(void);

#define EGL_OPENGL_ES_API 0x30A0
#define GL_VERTEX_SHADER 0x8B31
#define GL_FRAGMENT_SHADER 0x8B30
#define GL_ARRAY_BUFFER 0x8892
#define GL_UNIFORM_BUFFER 0x8A11
#define GL_STATIC_DRAW 0x88E4
#define GL_FLOAT 0x1406
#define GL_TRIANGLES 0x0004
#define GL_COLOR_BUFFER_BIT 0x4000
#define GL_RGBA 0x1908
#define GL_UNSIGNED_BYTE 0x1401
#define GL_NO_ERROR 0

#define W 64
#define H 64

static const char *VS =
    "attribute vec2 aPos;\n"
    "layout(std140, binding = 0) uniform Block { mat4 mvp; vec4 uColor; };\n"
    "void main() { gl_Position = mvp * vec4(aPos, 0.0, 1.0); }\n";

static const char *FS =
    "precision mediump float;\n"
    "layout(std140, binding = 0) uniform Block { mat4 mvp; vec4 uColor; };\n"
    "void main() { gl_FragColor = uColor; }\n";

static int near(unsigned char a, unsigned char b) {
    int d = (int)a - (int)b;
    return (d < 0 ? -d : d) <= 4;
}
static int is_green(const unsigned char *p) {
    return near(p[0], 0) && near(p[1], 255) && near(p[2], 0) && p[3] == 255;
}
static int is_blue(const unsigned char *p) {
    return near(p[0], 0) && near(p[1], 0) && near(p[2], 255) && p[3] == 255;
}

int main(void) {
    setbuf(stdout, NULL);

    EGLDisplay dpy = eglGetDisplay(0);
    if (!eglInitialize(dpy, 0, 0)) { fprintf(stderr, "eglInitialize failed\n"); return 1; }
    eglBindAPI(EGL_OPENGL_ES_API);
    EGLConfig cfg;
    EGLint num = 0;
    eglChooseConfig(dpy, 0, &cfg, 1, &num);
    EGLContext ctx = eglCreateContext(dpy, cfg, 0, 0);
    if (!ctx) { fprintf(stderr, "eglCreateContext failed\n"); return 1; }
    EGLSurface surf = eglCreatePbufferSurface(dpy, cfg, 0);
    if (!surf) { fprintf(stderr, "eglCreatePbufferSurface failed\n"); return 1; }
    if (!eglMakeCurrent(dpy, surf, surf, ctx)) { fprintf(stderr, "eglMakeCurrent failed\n"); return 1; }
    glViewport(0, 0, W, H);

    GLuint vs = glCreateShader(GL_VERTEX_SHADER);
    glShaderSource(vs, 1, &VS, 0);
    glCompileShader(vs);
    GLuint fs = glCreateShader(GL_FRAGMENT_SHADER);
    glShaderSource(fs, 1, &FS, 0);
    glCompileShader(fs);
    GLuint prog = glCreateProgram();
    glAttachShader(prog, vs);
    glAttachShader(prog, fs);
    glLinkProgram(prog);
    glUseProgram(prog);

    /* Full-NDC quad [-1,1]^2 as two triangles; the MVP relocates + shrinks it into a known rect. */
    float verts[12] = {
        -1.0f, -1.0f,  1.0f, -1.0f,  -1.0f, 1.0f,
         1.0f, -1.0f,  1.0f,  1.0f,  -1.0f, 1.0f,
    };
    GLuint vbo;
    glGenBuffers(1, &vbo);
    glBindBuffer(GL_ARRAY_BUFFER, vbo);
    glBufferData(GL_ARRAY_BUFFER, sizeof(verts), verts, GL_STATIC_DRAW);
    GLint aPos = glGetAttribLocation(prog, "aPos");
    if (aPos < 0) aPos = 0;
    glVertexAttribPointer((GLuint)aPos, 2, GL_FLOAT, 0, 8, 0);
    glEnableVertexAttribArray((GLuint)aPos);

    /* std140 block: mat4 mvp @0 (column-major) + vec4 uColor @64. mvp: x' = 0.5x + 0.25, y' = 0.5y.
     * -> quad NDC x in [-0.25,0.75] (cols [24,56)), y in [-0.5,0.5] (rows [16,48)). uColor = GREEN. */
    float block[20] = {
        /* mvp, column-major */
        0.5f,  0.0f, 0.0f, 0.0f, /* col0 */
        0.0f,  0.5f, 0.0f, 0.0f, /* col1 */
        0.0f,  0.0f, 1.0f, 0.0f, /* col2 */
        0.25f, 0.0f, 0.0f, 1.0f, /* col3: translate +0.25 in x */
        /* uColor = GREEN */
        0.0f, 1.0f, 0.0f, 1.0f,
    };
    GLuint ubo;
    glGenBuffers(1, &ubo);
    glBindBuffer(GL_UNIFORM_BUFFER, ubo);
    glBufferData(GL_UNIFORM_BUFFER, sizeof(block), block, GL_STATIC_DRAW);
    GLuint bidx = glGetUniformBlockIndex(prog, "Block");
    glUniformBlockBinding(prog, bidx, 0);
    glBindBufferBase(GL_UNIFORM_BUFFER, 0, ubo);

    glClearColor(0.0f, 0.0f, 1.0f, 1.0f); /* blue clear */
    glClear(GL_COLOR_BUFFER_BIT);
    glDrawArrays(GL_TRIANGLES, 0, 6);

    unsigned char *px = malloc(W * H * 4);
    memset(px, 0xAB, W * H * 4);
    glReadPixels(0, 0, W, H, GL_RGBA, GL_UNSIGNED_BYTE, px);
    GLenum e = glGetError();
    if (e != GL_NO_ERROR) { printf("GL_UBOXFORM_READBACK_FAILED err=0x%x\n", e); free(px); return 2; }

#define AT(x, y) (&px[(((y) * W) + (x)) * 4])
    unsigned char *hit = AT(40, 32);   /* inside the rect  -> GREEN */
    unsigned char *ml = AT(12, 32);    /* left of x'=-0.25 -> BLUE  */
    unsigned char *mr = AT(58, 32);    /* right of x'=0.75 -> BLUE  */
    unsigned char *mtop = AT(40, 6);   /* above the rect   -> BLUE  */
    unsigned char *mbot = AT(40, 58);  /* below the rect   -> BLUE  */
    printf("GL_UBOXFORM_HIT: %u %u %u %u\n", hit[0], hit[1], hit[2], hit[3]);
    printf("GL_UBOXFORM_ML: %u %u %u %u\n", ml[0], ml[1], ml[2], ml[3]);
    printf("GL_UBOXFORM_MR: %u %u %u %u\n", mr[0], mr[1], mr[2], mr[3]);

    int ok = is_green(hit) && is_blue(ml) && is_blue(mr) && is_blue(mtop) && is_blue(mbot);
    if (ok) {
        printf("GL_UBOXFORM_OK\n");
        free(px);
        return 0;
    }
    printf("GL_UBOXFORM_WRONG hit=(%u %u %u %u) ml=(%u %u %u %u) mr=(%u %u %u %u) — the std140 MVP block did "
           "not place the quad at the expected rect (cols 24..56, rows 16..48)\n",
           hit[0], hit[1], hit[2], hit[3], ml[0], ml[1], ml[2], ml[3], mr[0], mr[1], mr[2], mr[3]);
    free(px);
    return 3;
}

File: src/apps/husklet/tests/gl_ubo_transform.rs
//! REAL GL UBO TRANSFORM — a real EGL + GLES3 offscreen program whose MVP lives in a std140 uniform BLOCK
//! (set via glGenBuffers + glBufferData(std140) + glBindBufferBase + glUniformBlockBinding), rasterized on
//! lavapipe. The EXACT-RECT guard for UBO-block routing: the MVP both SCALES and TRANSLATES a full-NDC quad
//! into a KNOWN off-center rectangle (NDC x in [-0.25,0.75], y in [-0.5,0.5] -> pixel cols [24,56) x rows
//! [16,48) on a 64x64 target), filled GREEN from the block's uColor over a BLUE clear.
//!
//! This directly verifies the shim binds the `glBindBufferBase`d UBO buffer's std140 bytes at IR binding 0
//! (not a synthesized flat/zero block): a wrong route collapses the geometry (mvp=0 -> nothing) or misplaces
//! it (identity -> full frame), moving at least one of the sampled pixels off its expected color. We assert
//! the exact rect from BOTH the guest's own glReadPixels and an independent host-side read, and write the
//! host readback to /tmp/hl-demo/gl_ubo_transform.png.

use std::process::Command;

mod common;
use common::wgpu::WgpuExecutorServer;
use common::{bgra_to_rgba, demo_png_dir, staged_dir, write_png};

const W: u32 = 64;
const H: u32 = 64;

#[test]
fn real_gles3_ubo_mvp_places_quad_at_exact_rect_on_lavapipe() {
    let gl_dir = staged_dir("gl");
    assert!(
        gl_dir.join("libEGL.so.1").exists(),
        "staged libEGL.so.1 missing at {gl_dir:?}"
    );
    assert!(
        gl_dir.join("libGLESv2.so.2").exists(),
        "staged libGLESv2.so.2 missing at {gl_dir:?}"
    );

    let manifest = env!("CARGO_MANIFEST_DIR");
    let out_dir = std::env::temp_dir().join(format!("hl-demo-gluboxf-{}", std::process::id()));
    std::fs::create_dir_all(&out_dir).unwrap();
    let bin = out_dir.join("gl_ubo_transform");

    let compile = Command::new("gcc")
        .arg(format!(
            "{manifest}/../../surface/hl-gl/tests/fixtures/gl_ubo_transform.c"
        ))
        .arg(format!("-L{}", gl_dir.display()))
        .args(["-lEGL", "-lGLESv2"])
        .arg("-o")
        .arg(&bin)
        .output()
        .expect("spawn gcc");
    assert!(
        compile.status.success(),
        "gcc failed:\n{}",
        String::from_utf8_lossy(&compile.stderr)
    );

    let exec = WgpuExecutorServer::start("gluboxf");
    let adapter = exec.adapter_name();
    assert!(
        adapter.to_lowercase().contains("llvmpipe") || adapter.to_lowercase().contains("lavapipe"),
        "must rasterize on the software Vulkan device, got {adapter:?}"
    );

    let run = Command::new(&bin)
        .env("LD_LIBRARY_PATH", &gl_dir)
        .env("EGL_PLATFORM", "surfaceless")
        .env("HL_GPU_EXEC", exec.sock())
        .env("HL_GL_SURFACE_W", W.to_string())
        .env("HL_GL_SURFACE_H", H.to_string())
        .output()
        .expect("spawn gl_ubo_transform");
    let stdout = String::from_utf8_lossy(&run.stdout);
    let stderr = String::from_utf8_lossy(&run.stderr);
    eprintln!("--- gl_ubo_transform stdout ---\n{stdout}\n--- stderr ---\n{stderr}");

    assert!(
        run.status.success(),
        "guest exited non-zero (code {:?})",
        run.status.code()
    );
    assert!(
        stdout.contains("GL_UBOXFORM_OK"),
        "guest asserted the quad at the exact rect over the socket"
    );
    assert!(
        exec.submit_count() > 0,
        "guest submitted batches to the host executor"
    );

    // ---- Independent host-side read + PNG. Target is Bgra8Unorm (native B,G,R,A). GREEN (0,255,0) ->
    // BGRA (0,255,0); the BLUE clear (0,0,255) -> BGRA (255,0,0).
    let cap = exec.captured();
    let px = cap
        .rgba8_texture(W, H)
        .expect("no 64x64 target captured off the host executor");
    let near = |a: u8, b: u8| (a as i32 - b as i32).abs() <= 4;
    let at = |x: u32, y: u32| {
        let i = ((y * W + x) * 4) as usize;
        (px[i], px[i + 1], px[i + 2], px[i + 3]) // (B,G,R,A)
    };
    let is_green = |x, y| {
        let (b, g, r, a) = at(x, y);
        near(r, 0) && near(g, 255) && near(b, 0) && a == 255
    };
    let is_blue = |x, y| {
        let (b, g, r, a) = at(x, y);
        near(r, 0) && near(g, 0) && near(b, 255) && a == 255
    };

    write_png(
        &demo_png_dir().join("gl_ubo_transform.png"),
        W,
        H,
        &bgra_to_rgba(px),
    );

    assert!(
        is_green(40, 32),
        "rect interior must be GREEN, got {:?}",
        at(40, 32)
    );
    assert!(
        is_blue(12, 32),
        "left of the rect must be the BLUE clear, got {:?}",
        at(12, 32)
    );
    assert!(
        is_blue(58, 32),
        "right of the rect must be the BLUE clear, got {:?}",
        at(58, 32)
    );
    assert!(
        is_blue(40, 6),
        "above the rect must be the BLUE clear, got {:?}",
        at(40, 6)
    );
    assert!(
        is_blue(40, 58),
        "below the rect must be the BLUE clear, got {:?}",
        at(40, 58)
    );

    // The transformed GREEN fill must cover roughly the rect's area (32 cols x 32 rows = 1024 px of 4096),
    // proving a real scale+translate (not full-frame identity, not a collapsed zero matrix).
    let green = px
        .chunks_exact(4)
        .filter(|t| near(t[2], 0) && near(t[1], 255) && near(t[0], 0) && t[3] == 255)
        .count();
    assert!(
        (700..1400).contains(&green),
        "the transformed GREEN quad must cover ~1024 px (the exact rect), got {green}"
    );

    let _ = std::fs::remove_dir_all(&out_dir);
}
